retrybp/filters.go

```diff
@@ -0,0 +1,134 @@
+package retrybp
+
+import (
+	"context"
+	"errors"
+	"net"
+
+	retry "github.com/avast/retry-go"
+
+	"github.com/reddit/baseplate.go/clientpool"
+)
+
+var (
+	// DefaultFilterDecision is the default decision returned by Filters at the end
+	// of the FilterChain.
+	DefaultFilterDecision = false
+)
+
+// Filters returns a `retry.RetryIf` function that checks the error against the
+// given filters and returns either the decision reached by a filter or the
+// DefaultFilterDecision.
+//
+// You should not use this with any other retry.RetryIf options as one will
+// override the other.
+func Filters(filters ...RetryFilter) retry.Option {
+	fallbackValue := DefaultFilterDecision
+	next := &FilterChain{filter: func(_ error, _ *FilterChain) bool {
+		return fallbackValue
+	}}
+	for i := len(filters) - 1; i >= 0; i-- {
+		current := &FilterChain{
+			filter: filters[i],
+			next:   next,
+		}
+		next = current
+	}
+	chain := &FilterChain{next: next}
+	return retry.RetryIf(chain.Next)
+}
+
+// FilterChain is an ordered chain of RetryFilters that is used by Filter to
+// determine if a function should be retried based on the error it returns.
+//
+// A FilterChain is passed to RetryFilter functions and if the filter cannot make
+// a decison about whether to retry or not, it should instead return chain.Next
+// so the next filter in the chain can check the error.
+type FilterChain struct {
+	filter RetryFilter
+	next   *FilterChain
+}
+
+// Next returns the decision of the next RetryFilter in the FilterChain.
+func (c *FilterChain) Next(err error) bool {
+	if c.next.filter == nil || c.next == nil || c == nil {
+		return DefaultFilterDecision
+	}
+	return c.next.filter(err, c.next)
+}
+
+// RetryFilter is a function that is passed an error and attempts to determine
+// if the request should be retried given the error.
+//
+// RetryFilters should only implement a single check that will generally
+// determine whether the request should be retried or not.  If it cannot make
+// that decision on it's own, it should call the next filter in the given
+// FilterChain using chain.Next.
+//
+// If a filter is doing more than that, there's a good chance that it is doing
+// too much.
+type RetryFilter func(err error, chain *FilterChain) bool
```

Comment:
Is there a reason we need different types between `RetryFilter` and `FilterChain`? why can't it just be:

```
type RetryFilter func(err error, next RetryFilter) bool
```

We just need a termination implementation that ignore next, like:

```
func finalFilter(_ error, _ RetryFilter) bool {
  return DefaultFilterDecision
}
```

right?

Reply:
That approach doesn't quite work because you need to pass the next next into next, the Chain exists to hold that so you don't need it in `RetryFilter`.

Reply:
i.e.:

```
func noDecision(err error, next RetryFilter) bool {
  return next(err, ?????)
```

Reply:
nifty side benefit, `FilterChain.Next` is the same signature that `retry.RetryIf` wants so we can pass it right on into that.